README.md
SVG-TRANSLATION-MILL
--------------------
Licensed under: GPLv3

by Toni Serna <serna.toni@gmail.com>

Twitter: @toniserna

More info (SPANISH): http://cos.as/35l

The scenario
------------

If you come across an interesting svg (Scalable Vector Graphics), be a mind map, an infographic, the time line of events of "Pulp Fiction" or whatever, you will probably find  interesting to have it translated into a different language.

Without any help the process will be necessarily done from an svg editor like Inkscape on a copy of the original diagram. You will have to manually edit all the text boxes one by one and translate all the strings directly on the graphic.

This scenario worsens significantly if you are going to manage multiple translated versions of a graphic and need to contact a translator in order to get the job done. First of all you will need that the translator has an svg-compatible editor installed, then he or she must know the basics of editing vector graphics with such tool and besides you will probably have to check the final appearance of the graphic because the translator may have altered its layout.

The proposed solution
---------------------

Svg-translation-mill is a set of two bash scripts (svg2po.sh & po2svg.sh) that allow automatically generate a "po-like" file from a .svg, translate that file with a plain text editor and finally merge it back also automatically generating a new translated version of the original svg.

Installation
------------
  # You'll need pcregrep, a multiline capable grep-like command
  
sudo apt-get install pcregrep

  # Change with cd to your preferred folder and download there svg2po.sh and po2svg.sh
  
wget https://github.com/toniserna/svg-translation-mill/blob/master/po2svg.sh https://github.com/toniserna/svg-translation-mill/blob/master/svg2po.sh

  # Mark these scripts as executable
  
chmod a+x svg2po.sh po2svg.sh

  # Work as described in the next section of this README

The new workflow
----------------

Let's supose that you are the author of the svg file.

  1.  You will compose your graphic (let's name it SVG_FILE) using Inkscape or any other editor.  If any translation is needed, it includes different texts in your language of reference. If you are interested in facilitating contacting different translators consider to develop it at least an English version. When designing, keep texts as texts, do not convert them into paths or curves. It may be also interesting to export a bitmap version (.png, .jpg, .tiff etc.) of your SVG_FILE.

  2. From the command line and in the same folder wher you downloaded the two scripts, launch the first part of this mill:

     ./svg2po.sh SVG_FILE

   This will produce a file called a SVG_FILE.po (same folder of the original file). This file contains all the strings from the text boxes in your SVG_FILE in a format like this:

	msgid "Original text string 1"

	msgstr "-"


	msgid "Original text string 2"

	msgstr "-"


	msgid "Original text string 3"

	msgstr "-"

	...

  3. Send your translator the bitmap version of your SVG_FILE along with SVG_FILE.po

  4. Looking at the bitmap, the translator will be able to edit the SVG_FILE.po with any text editor. Other .po translation tools may be used (for example Virtaal) althoug some of them break the appropriate format of the file. An ordinay plain text editor will do the job.

  5. The translator just needs to send you back the SVG_FILE.po that must include all the translations. A translated file will look like this:

	msgid "Original text string 1"

	msgstr "Translated text string 1"


	msgid "Original text string 2"

	msgstr "Translated text string 2"


	msgid "Original text string 3"

	msgstr "Translated text string 3"

	...

  6. Time to start the second set of gears of this mill:

     ./po2svg.sh SVG_FILE

     After that you will get a file called SVG_FILE_translated.svg in the same folder of the original one.

  7. Open your new SVG_FILE_translated.svg with your Inkscape. You will have to manually adjust the line breaks in every text box and rearrange the position of the text boxes. This is because the SVG format splits every line of a text box in a different block <tspan>A line of text</tspan> every one placed automatically at its own coordinates. Once translated the lenght of strings may be completely different from original and thus is nearly impossible to automatically split the translated text into different lines. It has to be done manually.

  8. You are done.

Drawing for translation hints
-----------------------------
  * When designing an SVG graphic which has to be translated into several languages, try to keep every sentence which should be translated in the same text box.

  * Do not mix in the same text box different sentences.

Known limitations
-----------------
 * KL1: When using multiline text boxes, the first line of every text box in the graphic should be unique, as it is going to double as message identifier. In fact they do not only need to be unique, if one of the firsts strings of any text block is included as a starting substring of the first line of another text box, the po2svg.sh script won't be able to integrate the translations properly.

As an example of a problematic situation, imagine that in your svg you have a text box with the following two lines:

     NO
    ENTRY

You will have a problem if you have another text box like this:

    NO WAY OUT

Because "NO" (first text box) is a starting substring of "NO WAY OUT", po2svg.sh will use "NO" and "NO WAY OUT" as identificating strings.

It is free software
-------------------
Please feel free to use, study, comment, suggest, improve and share as you want as it is Free Software n the terms defined by GPLv3. Please keep me informed and credited.

No Freedom no Fun.


CHANGELOG.md
Changelog
---------
* 2012-06-03 v1.0
    - First release
    - Several improvements made at README.md
* 2012-06-04
    - TO_DO.md & this CHANGELOG.md added
    - Several improvements made at README.md and TO_DO.md

svg2po.sh
#!/bin/bash
#
# Usage:
#     svg2po.sh SVG_FILE
#
# Licensed under: GPLv3 (full text at LICENSE.md)
#
# Author: Toni Serna <serna.toni@gmail.com>
#
# Version/Date(YYYY-MM-DD): 1.0/2012-06-03
#
# Dependencies: pcregrep, po2svg.sh
#
# Recommended: inkscape
#
# Optional: virtaal
#
# Description:
#     Extracts the literal texts included in <text> ... </text> blocks from SVG_FILE.
# The result is sored in a file called SVG_FILE.po in the following format for each text block:
#
#     <blank line>
#     msgid "Literal text"
#     msgstr "-"
#
# In SVG format a multiline text bolck is split in multiple <tspan>text</tspan> blocks, one per line.
# This script concatenates in a single string all the <tspan>text</tspan> blocks in any <text> .. </text> blocks.
#
# Changelog (YYYY-MM-DD[:Version]):
# 
#     2012-06-03:1.0
#         First release
#

#----------------------------------------------------------
# Input SVG_FILE validation tests
SVG_FILE=$1

[ ! "$SVG_FILE" ] && echo -e "ERROR. Missing SVG_FILE\nUsage:\n    svg2po.sh SVG_FILE" >&2 && exit 1
if ! file -b "$SVG_FILE" | grep SVG >/dev/null
then
    echo -e "ERROR. This script is only able to process\nSVG (Scalable Vector Graphic) files" >&2
    exit 2
fi

# The hard core
sed -E -n '1h;1!H;${;g;s/<tspan[^>]*>/ /g;p;}' "$SVG_FILE" | sed 's/<\/tspan>//g' | pcregrep -Mo '<text[^>]*>.*<\/text>' | sed -E -n '1h;1!H;${;g;s/<text[^>]*> /\nmsgid "/g;p;}' | sed 's/<\/text>/"\nmsgstr "-"/' > "$SVG_FILE".po
#
# Regular expressions and chain of operations explained:
#
# 1. First "sed": It is able to operate with multiline regular expressions.
#    It searches all the <tspan [something]> tags and replaces them with a blank space.
# 2. Second "sed": In a multiline enabled expression it deletes all the </tspan> tags from the file.
#    At this point the different lines of every text box are put toghether at the same line with
#    an extra blank space at the beginning and a blank between every line.
# 3. pcregrep: It is a multiline-enabled version of the normal "grep", it returns only the parts of
#    the file that match the form: <text [something]>[Literal text]</text>
#    At this point we get only the part of the svg file in the form: <text [something]>Literal texts</tspan>
# 4. Third "sed": Another multiline expression to get rid of all the "<text [something]> " tags and
#    replace them for [newline]nmsgid ".
# 5. Fourth "sed": This removes the </text> tags and replaces them with a closing <">[newline]msgstr "-"
# 6. The result of this chain of transformations is stored in a file called SVG_FILE.po
#
NumMSGID=`cat "$SVG_FILE".po | wc -l`
NumMSGID=$[ $NumMSGID / 3 ] # We have 3 lines per text

[ $NumMSGID -eq 0 ] && "ERROR. No texts found in this file\nRemember that curves or bitmaps will not be processed" >&2 && rm "$SVG_FILE.po" && exit 3

clear
echo -e "Text extraction finished: $NumMSGID texts found.\nA file $SVG_FILE.po has been generated\n" >&2
echo -en "Would you like to start translating it now?\n(Your default editor will be opened)\n[y/n] " >&2
read -n1 ANSWER
echo

if [ $ANSWER == 'y' -o $ANSWER == 'Y' ]
then
    ${VISUAL:-${FCEDIT:-${EDITOR:-vi}}} "$SVG_FILE".po
else
    exit 0
fi

NumMSGSTR=`grep -c '^msgstr \"..\+\"' "$SVG_FILE".po`
if [ $NumMSGSTR -eq $NumMSGID ]
then
    clear
    echo -en "Finished translating?\n(If finished the new texts will be integrated in a new SVG file)\n[y/n] ">&2
    read -n1 ANSWER
    echo

    if [ $ANSWER == 'y' -o $ANSWER == 'Y' ]
    then
        #If svg-translation-mill is installed in a directory in the PATH,
        #the following call must not be preceeded by ./
        ./po2svg.sh "$SVG_FILE"
    else
        echo -e "When translation is ready remember to execute:\n    po2svg.sh $SVG_FILE" >&2
        exit 0
    fi
fi


po2svg.sh
#!/bin/bash
#
# Usage:
#     po2svg.sh SVG_FILE
#
# Licensed under: GPLv3 (full text at LICENSE.md)
#
# Author: Toni Serna <serna.toni@gmail.com>
#
# Version/Date(YYYY-MM-DD): 1.0/2012-06-03
#
# Dependencies: pcregrep
#
# Recommended: inkscape
#
# Optional: virtaal
#
# Description:
#     Integrates the texts translated at SVG_FILE.po back to SVG_FILE and generates a new version
# called SVG_FILE_translated.svg
#     SVG_FILE.po must have the following exact format for every string (as generated svg2po SVG_FILE):
#
#     <blank line>
#     msgid "Original string"
#     msgstr "Translated string"
#
#     In SVG format a multiline text bolck is split in multiple <tspan>text</tspan> blocks, one per line.
# This script integrates the translated strings as a one-line text block. Further manual editing required.
#
# Changelog (YYYY-MM-DD[:Version]):
# 
#     2012-06-03:1.0
#         First release
#

#----------------------------------------------------------
# Input SVG_FILE validation tests
SVG_FILE=$1

[ ! "$SVG_FILE" ] && echo -e "ERROR. Missing SVG_FILE\nUsage:\n    svg2po.sh SVG_FILE" >&2 && exit 1
if ! file -b "$SVG_FILE" | grep SVG >/dev/null
then
    echo -e "ERROR. This script is only able to process\nSVG (Scalable Vector Graphic) files" >&2
    exit 2
fi

# The whole file text is stored into the FILE variable
FILE=`cat "$SVG_FILE"`

# The following chain of operations extracts the first line of every multiline text blocks
pcregrep -Mo '<text[^>]*>([^<]*<tspan[^>]*>[^<]*<\/tspan>)*([^<]*<tspan[^>]*>)*[^<]*<\/text>' $SVG_FILE| sed -E -n '1h;1!H;${;g;s/<text[^>]*>[^<]*<tspan[^>]*>([^<]*)<\/tspan>([^<]*<tspan[^>]*>[^<]*<\/tspan>)*([^<]*<tspan[^>]*>)*[^<]*<\/text>/\1/g;p;}' | sort > /tmp/tempfile.$$
#
# Regular expressions and chain of operations explained:
#
# 1. "pcregrep": It is able to operate with multiline regular expressions.
#    It reduces the FILE to the <text [something]>...</text> tags. In the middle of these two tags
# a series of one(0) or more <tspan [something]>One-line string</tspan> will be found.
#    Notice that before the final </text> sometimes I found a non closed <tspan [something]> tag.
#    In order to make this seach mor robust and fail-proof I expected extra characters (possibly blanks
# or tabs) before opening any <tag>.
# 2. "sed": It is a "remembering" and multiline-enabled regular expression. Following the same schema
# explained before it "remembers" only the first line of any multi-line text blocks:
# the subregexp between "(" and ")". Then it reduces the whole <text>...</text> blocks to this remembered
# string.
#    Important: all of this srtings must be unique and none of them must be a substring found at the
# beginning of any of the rest of strings. This may require careful designing of the texts.
# 3. These texts are alphabetically sorted and stored in a temporary file
#

clear
NumMSGID=`cat /tmp/tempfile.$$ | wc -l`

[ $NumMSGID -eq 0 ] && "ERROR. No texts found in this file\nRemember that curves or bitmaps will not be processed" >&2 && rm "$1.po" && exit 3

echo -e "$NumMSGID strings found\nWould you like to check these identifying strings?\n(Your default editor will be opened)\n\nThey must be unique and not included at the beginning of any other" >&2
echo -en "[y/n] " >&2
read -n1 ANSWER
echo

if [ $ANSWER == 'y' -o $ANSWER == 'Y' ]
then
    ${VISUAL:-${FCEDIT:-${EDITOR:-vi}}} /tmp/tempfile.$$
fi

# The identifying strings are read from tempfile one by one into MSGID
while read MSGID
do
    # To avoid the special meaning in regexps of / and *, they are prefixed with \
    MSGID=`echo "$MSGID" | sed 's:\/:\\\/:g' | sed 's/\*/\\\*/g'`
    # MSGSTR will store the proper translation of the sentence which starts with MSGID at SVG_FILE.po
    # It is caught by reading the following line to the one which starts as "msgid "Identifying sting
    # The formating extras like -msgstr - and two " are removed
    MSGSTR=`grep -A 1 "msgid \"$MSGID" "$SVG_FILE".po | tail -1 | sed 's/^msgstr "//' | sed 's/"$//'`
    # Again, to avoid the special meaning in regexps of / and * ...
    MSGSTR=`echo "$MSGSTR" | sed 's:\/:\\\/:g' | sed 's/\*/\\\*/g'`

    # This performs the magic. In the entire text of the SVG_FILE it searches and replaces
    # Searches what?: the supposedly only <tspan></tspan> block which has the MSGID string
    # Replaced by what?: The whole translation for the text block as the first <tspan></tspan> block
    # at the corresponding <text></text>
    # And what about the second multiline "sed": It keeps the whole <text> block and the first
    # <tspan>Translated string</tspan> the rest of <tspan></tspan> int the same <text> are removed.
    # Note that a (or more) possible not closed final <tspan> tags are taken in consideration.
    # Also, in order to respond to extra blanks or tabs before every tag, extra [^<]* subexpresions
    # are included
    FILE=`echo "$FILE" | sed "s/>$MSGID[^<]*<\/tspan>/>$MSGSTR<\/tspan>/" | sed -E -n '1h;1!H;${;g;s/(<text[^>]*>[^<]*<tspan[^>]*>[^<]*<\/tspan>)([^<]*<tspan[^>]*>[^<]*<\/tspan>)*([^<]*<tspan[^>]*>)*[^<]*<\/text>/\1<\/text>/g;p;}'`

done < /tmp/tempfile.$$ # Every read whithin this loop will read from a line from the file

# The final result is flushed from memory to $SVG_FILE_translated.svg
echo "$FILE" > "${SVG_FILE}_translated.svg"
